packages/eslint-plugin/src/rules/no-misused-promises.ts

```diff
@@ -137,6 +158,164 @@ export default util.createRule<Options, 'conditional' | 'voidReturn'>({
       }
     }
 
+    function checkAssignments(node: TSESTree.AssignmentExpression): void {
+      const tsNode = parserServices.esTreeNodeToTSNodeMap.get(node);
+      const varType = checker.getTypeAtLocation(tsNode.left);
+      if (!isVoidReturningFunctionType(checker, tsNode.left, varType)) {
+        return;
+      }
+
+      if (returnsThenable(checker, tsNode.right)) {
+        context.report({
+          messageId: 'voidReturnVariable',
+          node: node.right,
+        });
+      }
+    }
+
+    function checkVariableDeclaration(node: TSESTree.VariableDeclarator): void {
+      const tsNode = parserServices.esTreeNodeToTSNodeMap.get(node);
+      if (tsNode.initializer === undefined || node.init === null) {
+        return;
+      }
+      const varType = checker.getTypeAtLocation(tsNode.name);
+      if (!isVoidReturningFunctionType(checker, tsNode.initializer, varType)) {
+        return;
+      }
+
+      if (returnsThenable(checker, tsNode.initializer)) {
+        context.report({
+          messageId: 'voidReturnVariable',
+          node: node.init,
+        });
+      }
+    }
+
+    function checkProperties(node: TSESTree.Property): void {
+      const tsNode = parserServices.esTreeNodeToTSNodeMap.get(node);
+      if (ts.isPropertyAssignment(tsNode)) {
+        const contextualType = checker.getContextualType(tsNode.initializer);
+        if (contextualType === undefined) {
+          return;
+        }
+        if (
+          !isVoidReturningFunctionType(
+            checker,
+            tsNode.initializer,
+            contextualType,
+          )
+        ) {
+          return;
+        }
+        if (returnsThenable(checker, tsNode.initializer)) {
```

Comment:
Just nitpicking here, but there are a lot of consecutive `if` statements with `return`s in this PR. You could reduce the lines of code and IMO improve readability a bit by merging them:

```suggestion
          contextualType !== undefined &&
          isVoidReturningFunctionType(
            checker,
            tsNode.initializer,
            contextualType,
          ) &&
          returnsThenable(checker, tsNode.initializer)
        ) {
```

What do you think? If you disagree, no worries, not a strong blocker IMO 🙂

Reply:
Agreed :smiley: fixed!